File: Embassat/View/Info/CADEMInfoViewController.m
//
//  CADEMInfoViewController.m
//  Embassat
//
//  Created by Joan Romano on 30/03/14.
//  Copyright (c) 2014 Crows And Dogs. All rights reserved.
//

#import "CADEMInfoViewController.h"

#import "CADEMInfoCollectionViewCell.h"
#import "CADEMInfoHeaderView.h"
#import "CADArrayDataSource.h"

#import "CADEMInfoViewModel.h"

@interface CADEMInfoViewController ()

@property (nonatomic, weak) IBOutlet UICollectionView *infoCollectionView;
@property (nonatomic, strong) CADEMInfoCollectionViewCell *prototypeCell;

@property (nonatomic, strong) CADArrayDataSource *dataSource;

@end

@implementation CADEMInfoViewController

- (instancetype)init
{
    if (self = [super init])
    {
        self.title = @"Info";
    }
    
    return self;
}

- (void)viewDidLoad
{
    [super viewDidLoad];
    
    self.infoCollectionView.dataSource = self.dataSource;
    [self.infoCollectionView registerNib:[UINib nibWithNibName:@"CADEMInfoCollectionViewCell" bundle:nil] forCellWithReuseIdentifier:CADCellIdentifier];
    [self.infoCollectionView registerNib:[UINib nibWithNibName:@"CADEMInfoHeaderView" bundle:nil] forSupplementaryViewOfKind:UICollectionElementKindSectionHeader withReuseIdentifier:CADHeaderIdentifier];
    self.prototypeCell = [[[UINib nibWithNibName:@"CADEMInfoCollectionViewCell" bundle:nil] instantiateWithOwner:nil options:nil] firstObject];
}

- (CGSize)collectionView:(UICollectionView *)collectionView layout:(UICollectionViewLayout*)collectionViewLayout sizeForItemAtIndexPath:(NSIndexPath *)indexPath
{
    self.prototypeCell.body = [self.viewModel bodyAtIndexPath:indexPath];
    
    return [self.prototypeCell systemLayoutSizeFittingSize:UILayoutFittingCompressedSize];
}

#pragma mark - Lazy

- (CADArrayDataSource *)dataSource
{
    if (!_dataSource)
    {
        _dataSource = [[CADArrayDataSource alloc] initWithViewModel:self.viewModel configureCellBlock:^(CADEMInfoCollectionViewCell *cell, id indexPath) {
            cell.body = [self.viewModel bodyAtIndexPath:indexPath];
        } configureHeaderBlock:^(CADEMInfoHeaderView *header, id indexPath) {
            header.title = [self.viewModel titleAtIndexPath:indexPath];
            header.cover = [self.viewModel imageAtIndexPath:indexPath];
        }];
    }
    
    return _dataSource;
}

@end

File: Embassat/View/Info/CADEMInfoViewController.swift
//
//  CADEMInfoViewController.swift
//  Embassa't
//
//  Created by Joan Romano on 24/05/15.
//  Copyright (c) 2015 Crows And Dogs. All rights reserved.
//

import UIKit

public class CADEMInfoViewController: CADEMRootViewController {
    
    @IBOutlet weak var infoCollectionView: UICollectionView?
    let prototypeCell: CADEMInfoCollectionViewCell
    let dataSource: CADArrayDataSource
    let viewModel: CADEMInfoViewModel
    
    override init(nibName nibNameOrNil: String?, bundle nibBundle: NSBundle?) {
        
        let theViewModel = CADEMInfoViewModel()
        prototypeCell = UINib(nibName: "CADEMInfoCollectionViewCell", bundle: nil).instantiateWithOwner(nil, options: nil).first as! CADEMInfoCollectionViewCell
        dataSource =
            CADArrayDataSource(viewModel: theViewModel,
                configureCellBlock: { (cell: AnyObject!, indexPath: NSIndexPath) -> Void in
                    let theCell = cell as! CADEMInfoCollectionViewCell
            
                    theCell.body = theViewModel.bodyAtIndexPath(indexPath)
                },
                configureHeaderBlock: { (header: AnyObject!, indexPath: NSIndexPath) -> Void in
                    let theHeader = header as! CADEMInfoHeaderView
                
                    theHeader.title = theViewModel.titleAtIndexPath(indexPath)
                    theHeader.cover = theViewModel.imageAtIndexPath(indexPath)
            })
        viewModel = theViewModel
        
        super.init(nibName: nibNameOrNil, bundle: nibBundle)
    }
    
    required public convenience init(coder aDecoder: NSCoder) {
        self.init(nibName: nil, bundle: nil)
    }
    
    override public func viewDidLoad() {
        super.viewDidLoad()
        
        title = "Info"
        
        infoCollectionView?.dataSource = dataSource
        infoCollectionView?.registerNib(UINib(nibName: "CADEMInfoCollectionViewCell", bundle: nil), forCellWithReuseIdentifier: CADArrayDataSource.CADCellIdentifier)
        infoCollectionView?.registerNib(UINib(nibName: "CADEMInfoHeaderView", bundle: nil), forSupplementaryViewOfKind: UICollectionElementKindSectionHeader, withReuseIdentifier: CADArrayDataSource.CADHeaderIdentifier)
    }
    
    func collectionView(collectionView: UICollectionView, layout collectionViewLayout: UICollectionViewLayout, sizeForItemAtIndexPath indexPath: NSIndexPath) -> CGSize {
        prototypeCell.body = viewModel.bodyAtIndexPath(indexPath)
        
        return prototypeCell.systemLayoutSizeFittingSize(UILayoutFittingCompressedSize)
    }
}
